<h1 style='color:tomato'><b> Home Assignment 4 </b></h1>
<h2>Studio dell'Entanglement in sistemi a 2 Qubit</h2>

# <h2 style="color:Blue"> Import </h2>

In [11]:
from qiskit import QuantumCircuit
from qiskit_aer import Aer
import numpy as np

def exec(c,shots):
    backend = Aer.get_backend('statevector_simulator') #creo il simulatore
    job = backend.run(c, shots = shots,seed_simulator = 424242) #eseguo con un seed fissato (riproducibilà dei risultati)
    result = job.result() #ottengo i risultati
    counts = result.get_counts() #salvo in un dizionario
    print(c.draw()) #stampo il circuito
    print(counts) #stampo i risultati

# <h2 style="color:Blue"> Punto 1 </h2>

### a)
Si costruisca lo stato <br>
$$ \color{orange}{|\psi\rangle = \alpha|0\rangle + \beta|1\rangle}\color{black} \ \ \ \text{ con } \ |\alpha|^2 = 0.4 \text{ e } |\beta|^2=0.6 $$

per il $\color{orange}{\text{primo qubit}}$, e lo stato $\color{blue}{|{-}\rangle}$ per il $\color{blue}{\text{secondo qubit}}$.

Creiamo un circuito quantistico a 2 Qubit con i relativi registri classici per contenere le misure ottenute.

In [12]:
c1 = QuantumCircuit(2,2)

Inizializziamo il primo qubit $Q_0$ di modo che, quando misurato con l'operatore $X$, il $40\%$ delle volte collassa nell'autostato $|0\rangle$ e il $60\%$ in $|1\rangle$.

Supponiamo di avere un <b>set universale</b> composto solamente dalle porte:
- $X \rightarrow$ NOT quantistico in Base Canonica $B_0 = \{|0\rangle,|1\rangle\}$:<br>
    $$|0\rangle \leftrightarrow |1\rangle$$

- $H \rightarrow$ porta di Hadamard, crea sovrapposizione degli stati di $B_0$, in pratica converte gli autovalori di $Z$ negli autovalori di $X$ e viceversa:<br>
    $$|0\rangle \leftrightarrow |+\rangle \color{gray}{= \frac{|0\rangle+|1\rangle}{\sqrt2}}\color{black} \ \ \ \text{ e  } \ \ \ |1 \leftrightarrow |-\rangle \color{gray}{= \frac{|0-|1\rangle}{\sqrt2}} $$

- $\mathrm{U}_{\mathrm{ph}} (\gamma) \rightarrow$ Phase Gate, implementa un cambio di fase relativa su $|1\rangle$<br>
     $$|0\rangle \leftrightarrow |0\rangle \ \ \ \text{ e } \ \ \ |1\rangle \rightarrow e^{i\gamma}|1\rangle$$

<br>
<br>

Seguendo l'algoritmo di inizializzazione di un qubit visto a lezione, possiamo produrre uno stato<br>
$$|\psi\rangle = \cos \delta |0\rangle + e^{i \gamma} \sin \delta |1\rangle \ \ \ \text {con } \ \delta, \gamma \ \text{ arbitrari}$$

Partendo dallo stato $|0\rangle$, applico in ordine:
1. $H$
    $$ |0\rangle \ \ \ \overset H\longrightarrow \ \ \ \frac1{\sqrt2}\Big(|0\rangle+|1\rangle\Big)$$
2. $\mathrm{U}_{\mathrm{ph}}(-2\delta)$
    $$  \overset {\mathrm{U}_{\mathrm{ph}}(-2\delta)} \longrightarrow \ \ \ \frac1{\sqrt2}\Big(|0\rangle+e^{-2i\delta}|1\rangle\Big)$$
3. $H$
    $$
    \begin{matrix}
    \overset H\longrightarrow & \frac12\Big[ |0\rangle + |1\rangle + e^{-2i\delta}\big(|0\rangle - |1\rangle \big) \Big] \\ \ \\

    = & \frac12\Big[ \big(1+ e^{-2i\delta}\big)|0\rangle + \big(1- e^{-2i\delta}\big)|1\rangle \Big] \\ \ \\

    = & \frac{e^{-i\delta}}2\Big[ \big(e^{i\delta}+ e^{-i\delta}\big)|0\rangle + \big(e^{i\delta}- e^{-i\delta}\big)|1\rangle \Big]
    \\ \ \\

    = & \cos\delta|0\rangle + i\sin\delta|1\rangle
    \end{matrix}
    $$
4. $\mathrm{U}_{\mathrm{ph}}(-\pi/2)$
    $$  \overset {\mathrm{U}_{\mathrm{ph}}(-\frac\pi2)} \longrightarrow \ \ \  \cos\delta|0\rangle + \sin\delta|1\rangle $$
5. $\mathrm{U}_{\mathrm{ph}}(\gamma)$
    $$  \overset {\mathrm{U}_{\mathrm{ph}}(\gamma)} \longrightarrow \ \ \ \cos\delta|0\rangle + e^{i\gamma}\sin\delta|1\rangle $$

Per semplicitá NON utilizziamo una fase relativa, in quanto non necessaria poiché non viene presa in considerazione dalla misurazione di $X$; per questo motivo non eseguiremo gli ultimi 2 passi dell'algoritmo.

In particolare per il nostro stato, utilizziamo un valore di $\delta$ tale che $\cos(\delta)=\sqrt{0.4}$ (ovvero che la probabilitá di collassare in $|0\rangle$ sia $0.4$, come richiesto).

Nel nostro caso, scegliamo $\delta = 0.88$ radianti.

In [13]:
delta = np.arccos(np.sqrt(0.4))
c1.h(0)
c1.p(-2*delta,0)
c1.h(0)
#|Ψ> = cos(0.88)|0> + sin(0.88)|1>

Poiché utilizziamo operatori unari, possiamo rappresentarli sulla sfera di bloch come fossero rotazioni. <br>

Partendo da $|0\rangle$, osserviamo in sequenza $H$, $\mathrm{U}_{\mathrm{ph}}(-1.7722)$, $H$.


<img src="Q1_00_H.png" width=300>
<img src="Q1_01_U.png" width=300>
<img src="Q1_02_H.png" width=300>

Inizializzo anche Q2, semplicemente usando $X$, per trasformarlo da $|0\rangle$ in $|1\rangle$ e successivamente $H$ per traformarlo in $|-\rangle$

Ricordiamo $|-\rangle = \frac{1}{\sqrt{2}}(|0\rangle-|1\rangle)$

In [14]:
c1.x(1) #da |0> in |1>
c1.h(1) #da |1> in |->

Rappresentiamo anche l'inizializzazione del secondo Qubit (prima $X$ e poi $H$).

<img src="Q2_00_X.png" width=300>
<img src="Q2_01_H.png" width=300>

### b) 
Si faccia una misura del sistema; i risultati del primo e del secondo qubit sono correlati?

Misuro infine $X$, secondo la Teoria Della Misura misurando l'operatore Hermittiano $X$ (diagonale in Base Canonica $\{|0\rangle,|1\rangle\}$) su un qubit con stato nella forma <br>

$$|\psi\rangle=\cos (0.88) |0\rangle + \sin ( 0.88) |1\rangle$$

questo collassa nello stato :
* $|0\rangle$ (con relativo autovalore +1) con $ P = \cos^2(0.88) = 0.637^2 = 0.40576 \approx 0.4$ 
* $|1\rangle$ (con relativo autovalore -1) con $ P = \sin^2(0.88) = 0.770^2 = 0.5929 \approx 0.6$ 

Similmente, misurando $Q_1$, poiché in completa sovrapposizione di stati $|0\rangle, |1\rangle$, questo collasserá in $|0\rangle$ con $P=0.5$ e in $|1\rangle$ con la stessa probabilitá.

In [15]:
#MISURO
c1.barrier()
#le misure sono invertite (little vs big endian)
c1.measure(0, 1)
c1.measure(1, 0)

Poichè la misura è probabilistica (fatta eccezione per le misure direttamente sugli autostati dell' operatore Hermittiano, che in questo caso non si verificano), per ottenere risultati significativi eseguiamo 100.000 shots:

In [16]:
exec(c1,100000)

     ┌───┐┌────────────┐┌───┐ ░ ┌─┐   
q_0: ┤ H ├┤ P(-1.7722) ├┤ H ├─░─┤M├───
     ├───┤└───┬───┬────┘└───┘ ░ └╥┘┌─┐
q_1: ┤ X ├────┤ H ├───────────░──╫─┤M├
     └───┘    └───┘           ░  ║ └╥┘
c: 2/════════════════════════════╩══╩═
                                 1  0 
{'01': 20146, '10': 30036, '00': 19834, '11': 29984}


Mostriamo innanzitutto come possiamo descrivere lo stato $|\Phi\rangle$ del nostro sistema come un prodotto tensore tra gli stati di 2 qubit:

$$|\Phi\rangle = Q_0\otimes Q_1 $$

Questa osservazione ci permette di considerare i 2 qubit "separati", in quanto non c'é correlazione tra loro.
<br>
<br>
<br>
Mostriamo i risultati in forma matriciale

$$ \begin{array}{c|c}
  Q_1/Q_0   & Q_0 = |0\rangle & Q_0 = |1\rangle & \text{Sum} \\ \hline
  Q_1 = |0\rangle &  19.834 & 30.036 & \color{green}{\boxed{49.870}}\color{black}\\
  Q_1 = |1\rangle & 20.146 & 29.984 & \color{green}{\boxed{50.130}}\color{black}\\
  \text{Sum} & \color{orange}{\boxed{39.980}}\color{black} & \color{orange}{\boxed{60.020}}\color{black} & 100.000 \\
\end{array} $$

Osserviamo prima di tutto le colonne marginali, possiamo notare come: 
* per il qubit $Q_0$, come richiesto, le probabilitá siano $\color{orange}{\boxed{40}}\color{black}$ e $\color{orange}{\boxed{60}}\color{black}$;
* mentre per il qubit $Q_1$ siano $\color{green}{\boxed{50}}\color{black}/\color{green}{\boxed{50}}\color{black}$ confermando la perfetta sovrapposizione descritta nello stato $|-\rangle$.

Mostriamo ora che i 2 qubit <b>NON</b> sono correlati:
- Supponiamo ad esempio che $Q_0$ misuri $|0\rangle$, ovvero controlliamo i risultati solo sulla prima colonna; le probabilitá di misurare $Q_1$ restano $50$ e $50$.
- Allo stesso modo, se $Q_1$ misura $|1\rangle$, possiamo facilmente verificare come i valori sulla seconda colonna siano nuovamente distribuiti con probabilitá $50$ e $50$.
- Decidiamo ora di misurare prima $Q_1$: se otteniamo $|0\rangle$ (prima riga), poco sorprendentemente $Q_0$ otterà $|0\rangle$ e $|1\rangle$ con probabilitá rispettivamente $40$ e $60$ (lo stato é stato creato apposta in questo modo).
- Infine, se $Q_1$ collassa in $|1\rangle$, $Q_0$ collassa in $|0\rangle$ e $|1\rangle$ con le stesse probabilitá.


Ma cosa abbiamo osservato quindi? Che i 2 Qubit vivono in uno spazio "separato" e completamente slegato uno dall'altro. <br>
Questo lo abbiamo implicitamente giá osservatao al momento della "creazione" degli stati $|\psi\rangle$ e $|-\rangle$ : non ci siamo minimante preoccupati del numero di qubit presenti nello spazio o di che stati avessero, ma ognuno é elaborato "a se stante".<br>

Questa cosa la si puó notare anche nella rappresentazione grafica del circuito quantistico: ogni porta applicata é unaria, e quindi non abbiamo mai fatto interagire gli stati di 2 qubit diversi.

# <h2 style="color:blue"> Punto 2 </h2>

### a) 
Si generi lo stato di Bell 

$$ \color{orange}{|{\Phi^{+}\rangle} =\frac{1}{\sqrt{2}}(|{00}\rangle + |{11}\rangle)} $$

Mostriamo innanzitutto come questo stato <b>NON</b> possa essere descritto come prodotto tensore tra gli stati generico $\color{blue}{| v\rangle}$ e $\color{green}{| w\rangle}$ dei qubit $\color{blue}{Q_0}$, $\color{green}{Q_1}$:

$$ \color{blue}{| v\rangle = \alpha | 0\rangle + \beta | 1\rangle} $$

$$ \color{green}{| v\rangle = \gamma | 0\rangle + \delta | 1\rangle} $$

Lo dimostriamo supponendo che si possa <b>SEMPRE</b> descrivere lo stato di un sistema di N Qubit come prodotto tensore dei singoli presi separatamente:

$$ \begin{align*}
\color{orange}{|{\Phi}\rangle}\color{black} &= \color{blue}{|{v}\rangle}\color{black} \otimes \color{green}{|{w}\rangle}\color{black} \\
&= \color{blue}{(\alpha |{0}\rangle + \beta |{1}\rangle)}\color{black} \otimes \color{green}{(\gamma |{0}\rangle + \delta |{1}\rangle)}\color{black} \\
&= \color{blue}{\alpha} \color{green}{\gamma} \color{black}|{\color{blue}{0}\color{green}{0}}\color{black}\rangle + \color{blue}{\alpha} \color{green}{\delta}\color{black} |{\color{blue}{0}\color{green}{1}}\color{black}\rangle  + \color{blue}{\beta} \color{green}{\gamma}\color{black} |{\color{blue}{1}\color{green}{0}}\color{black}\rangle + \color{blue}{\beta} \color{green}{\delta}\color{black} |{\color{blue}{1}\color{green}{1}}\color{black}\rangle
\end{align*} $$

In particolare, nel caso di $\color{orange}{|{\Phi^{+}\rangle}}$ <br>
$$ \begin{align*}
\color{blue}{\alpha} \color{green}{\gamma} &= \frac{1}{\sqrt{2}} \\
\color{blue}{\alpha} \color{green}{\delta} &= 0 \\
\color{blue}{\beta} \color{green}{\gamma} &= 0 \\
\color{blue}{\beta} \color{green}{\delta} &= \frac{1}{\sqrt{2}}
\end{align*} $$

Che messi a sistema creano una <b style="color:red">CONTRADDIZIONE</b>.

Cosa possiamo dedurre questa dimostrazione? <br>
Possiamo affermare che il nostro sistema "descrive" un super-stato $ \color{orange}{|{\Phi^{+}\rangle}} $ che contiene piú "informazione" della somma dei 2 sotto-stati presi separatamente.

Per definizione gli stati di Bell sono 4 stati (tra loro ortogonali) che formano una base del super-spazio, e sono tutti vettori <b style="color:red">MASSIMAMENTE ENTANGLED</b>.


Produciamo ora lo stato $ \color{orange}{|{\Phi^{+}\rangle}} $ partendo dai singoli stati $\color{blue}{| 0\rangle}$, $\color{green}{| 0\rangle}$.

Applico in ordine:
1. $H_0 \rightarrow $ Porta di Hadamard, per creare sovrapposizione su $\color{blue}{Q_0}$.
    $$ \color{blue}{|0\rangle}\color{black} \otimes \color{green}{|0\rangle}\color{black} \ \ \ \overset {H_{0}} \longrightarrow \ \ \ \color{blue}{|{+}\rangle}\color{black} \otimes \color{green}{|0\rangle}\color{black} \ \ \ = \ \ \ \color{blue}{\frac{|0\rangle+|1\rangle}{\sqrt2}}\color{black} \otimes \color{green}{|0\rangle}\color{black} \ \ \  = \ \ \ \frac{|{\color{blue}{0}\color{green}{0}}\color{black}\rangle+|{\color{blue}{1}\color{green}{{0}}\color{black}\rangle}}{\sqrt2} $$

2. $C_{0}NOT_{1} \rightarrow $ Porta $\color{orange}{\text{"Controlled NOT"}}$, che trasforma $\color{green}{| 0\rangle}\color{black} \rightarrow \color{green}{| 1\rangle}$ ma solo se $\color{blue}{Q_0}$ = $\color{blue}{| 0\rangle}$.
    $$ \frac{|{\color{blue}{0}\color{green}{0}}\color{black}\rangle+|{\color{blue}{1}\color{green}{{0}}\color{black}\rangle}}{\sqrt2}  \ \ \ \overset {C_{0}NOT_{1}} \longrightarrow \ \ \ \frac{|{\color{blue}{0}\color{green}{0}}\color{black}\rangle+|{\color{blue}{1}\color{green}{1}}\color{black}\rangle}{\sqrt2} \ \ \ = \ \ \ \color{orange}{|{\Phi^{+}\rangle}}$$

É bene sottolineare come sia l'utilizzo della porta binaria $C_{0}NOT_{1}$ che ci permette di creare $\color{orange}{\text{Entanglement}}$: lo si puó notare dal circuito stesso, in cui notiamo come la porta, prendendo in considerazione piú Qubit, non ci permetta piú di poterli considerare "scollegati".

In [17]:
c2 = QuantumCircuit(2,2) 

c2.h(0)
c2.cx(0,1)

### b) 
Si faccia una misura del sistema; i risultati del primo e del secondo qubit sono correlati?

Completiamo allora il circuito con le misurazioni sui qubit $\color{blue}{Q_0}\color{black} , \color{green}{Q_1}$

In [18]:
#MISURO
c2.barrier()
#le misure sono invertite (little vs big endian)
c2.measure(0, 1)
c2.measure(1, 0)

Per lo stesso ragionamento fatto in precedenza, eseguiamo 100.000 shots.

In [19]:
exec(c2,100000)

     ┌───┐      ░ ┌─┐   
q_0: ┤ H ├──■───░─┤M├───
     └───┘┌─┴─┐ ░ └╥┘┌─┐
q_1: ─────┤ X ├─░──╫─┤M├
          └───┘ ░  ║ └╥┘
c: 2/══════════════╩══╩═
                   1  0 
{'00': 50050, '11': 49950}


Analogamente a prima, mostriamo i risultati all'interno di una matrice.

$$ \begin{array}{c|c}
  \color{green}{Q_1}\color{black}/\color{blue}{Q_0}\color{black}   & Q0 = |0\rangle & Q0 =|1\rangle & \text{Sum} \\ \hline
  Q1 = |0\rangle & \color{magenta}{\boxed{49950}} & 0 & 49950 \\
  Q1 = |1\rangle & 0 & \color{cyan}{\boxed{50050}} & 50050\\
  \text{Sum} & 49950 & 50050 & 100.000 \\
\end{array} $$

é immediato notare come la matrice sia <b style="color:powderblue">DIAGONALE</b>: questa caratteristica é tipica degli stati <b style="color:red">massimamente entangled</b>; <br>
ovvero, ci assicura che una volta fatta la misura su un Qubit, la misura su quello restante sará <b>DETERMINISTICA</b>, in quanto entrambi i Qubit saranno collassati.

Cerchiamo ora di capire in che senso "collassano" entrambi. <br>
Prendiamo il nostro stato 
$$ \color{orange}{|{\Phi^{+}\rangle}}\color{black} = \frac{1}{\sqrt2}\big(\color{blue}{|0\rangle}\color{black} \otimes \color{green}{|0\rangle}\color{black}+\color{blue}{|1\rangle}\color{black}\otimes \color{green}{|1\rangle}\color{black}\big) $$
  

Supponiamo che la misura su $\color{blue}{Q_0}$ restituisca 1, ovvero collassi nello stato $\color{blue}{|{0}\rangle}$.

$$ \frac{1}{\sqrt2}\big(\color{blue}{|0\rangle}\color{black} \otimes \color{green}{|0\rangle}\color{black}+\color{blue}{|1\rangle}\color{black}\otimes \color{green}{|1\rangle}\color{black}\big) \ \ \ \overset {Z_0} \longrightarrow \ \ \ \color{blue}{|0\rangle}\color{black} \otimes \color{green}{|0\rangle}\color{black}  \text{ con }  P = (\frac{1}{\sqrt2})^2 = 0.5$$

Dal punto di vista pratico, ci stiamo concentrando sulla prima riga della matrice, e abbiamo escluso la seconda.

Effettuiamo ora anche la misura su $\color{green}{Q_1}$ sullo stato ottenuto.

$$ \color{blue}{|0\rangle}\color{black} \otimes \color{green}{|0\rangle}\color{black} \ \ \ \overset {Z_1} \longrightarrow \ \ \ \color{blue}{|0\rangle}\color{black} \otimes \color{green}{|0\rangle}\color{black} \text{ con }  P = (1)^2 = 1 $$

Ovvero abbiamo uno stato $ \color{orange}{ \hat{|{\Phi^{+}\rangle}}} $ codificato da uno stato di $\color{green}{Q_1}$ che é giá un autostato della misura $\color{green}{Z_1}$.

E se la misura $\color{blue}{Z_0}$ restituisse -1? <br>
Allora in questo caso avremmo lo stato collassato in $ \color{blue}{|1\rangle}\color{black} \otimes \color{green}{|1\rangle} $, e le conclusioni precedenti su $\color{green}{Q_1}$ sono le stesse.

Abbiamo inoltre la conferma della probabilitá $50/50$ osservando i valori $\color{magenta}{\boxed{49950}} $/$ \color{cyan}{\boxed{50050}} $.

E se volessimo misurare prima $\color{green}{Z_1}$ e poi $\color{blue}{Z_0}$?<br>
I risultati sarebbero analoghi, e basterebbe considerare la matrice trasposta.

Ovviamente nel nostro caso, in cui la matrice risultante é diagonale, le marginali non servono, ma le indichiamo per completezza.



### c) 
Si ripetano i punti precedenti per lo stato di Bell 

$$ \color{orange}{|{\psi^{-}\rangle} =\frac{1}{\sqrt{2}}(|{01}\rangle - |{10}\rangle)} $$

Essendo questo un'altro dei 4 stati della <b style="color:red">Base di Bell</b>, valgono le stesse proprietá studiate in precedenza.

Produciamo ora lo stato $ \color{orange}{|{\psi^{-}\rangle}} $ partendo dai singoli stati $\color{blue}{| 0\rangle}$, $\color{green}{| 0\rangle}$.

Applico in ordine:
1. $H_0 \rightarrow $ Porta di Hadamard, per creare sovrapposizione su $\color{blue}{Q_0}$.
    $$ \color{blue}{|0\rangle}\color{black} \otimes \color{green}{|0\rangle}\color{black} \ \ \ \overset {H_{0}} \longrightarrow \ \ \ \color{blue}{|{+}\rangle}\color{black} \otimes \color{green}{|0\rangle}$$

2. $Z_0 \rightarrow $ Porta di Pauli, la usiamo per traformare  $\color{blue}{|{+}\rangle}\color{black} \rightarrow \color{blue}{|{-}\rangle}$
    $$ \color{blue}{|{+}\rangle}\color{black} \otimes \color{green}{|0\rangle}\color{black} \overset {Z_{0}} \longrightarrow \ \ \ \color{blue}{|{-}\rangle}\color{black} \otimes \color{green}{|0\rangle}\color{black} \ \ \ = \ \ \ \color{blue}{\frac{|0\rangle-|1\rangle}{\sqrt2}}\color{black} \otimes \color{green}{|0\rangle} $$

3. $X_1 \rightarrow $ Porta di Pauli, la usiamo per traformare  $\color{green}{|0\rangle}\color{black} \rightarrow \color{green}{|1\rangle}$ .
    $$ \color{blue}{\frac{|0\rangle-|1\rangle}{\sqrt2}}\color{black} \otimes \color{green}{|0\rangle}\color{black} \ \ \ \overset {X_1} \longrightarrow \ \ \ \color{blue}{\frac{|0\rangle-|1\rangle}{\sqrt2}} \color{black}\otimes \color{green}{|1\rangle}\color{black} \ \ \ = \ \ \ \frac{|{\color{blue}{0}\color{green}{1}}\color{black}\rangle-|{\color{blue}{1}\color{green}{1}}\rangle}{\sqrt2} $$

4. $C_{0}NOT_{1} \rightarrow $ Porta $\color{orange}{\text{"Controlled NOT"}}$, che trasforma $\color{green}{| 0\rangle}\color{black} \rightarrow \color{green}{| 1\rangle}$ ma solo se $\color{blue}{Q_0}$ = $\color{blue}{| 0\rangle}$.
    $$ \frac{|{\color{blue}{0}\color{green}{1}}\color{black}\rangle-|{\color{blue}{1}\color{green}{1}}\rangle}{\sqrt2}\color{black}  \ \ \ \overset {C_{0}NOT_{1}} \longrightarrow \ \ \ \frac{|{\color{blue}{0}\color{green}{1}}\color{black}\rangle-|{\color{blue}{1}\color{green}{0}}\rangle}{\sqrt2}\color{black} \ \ \ = \ \ \ \color{orange}{|{\Psi^{-}}\rangle} $$


In [20]:
c3 = QuantumCircuit(2,2) 

c3.h(0);c3.z(0)

c3.x(1)

c3.cx(0,1)

#MISURO
c3.barrier()
#le misure sono invertite (little  vs big endian)
c3.measure(0, 1)
c3.measure(1, 0)

exec(c3,100000)

     ┌───┐┌───┐      ░ ┌─┐   
q_0: ┤ H ├┤ Z ├──■───░─┤M├───
     ├───┤└───┘┌─┴─┐ ░ └╥┘┌─┐
q_1: ┤ X ├─────┤ X ├─░──╫─┤M├
     └───┘     └───┘ ░  ║ └╥┘
c: 2/═══════════════════╩══╩═
                        1  0 
{'01': 50050, '10': 49950}


Mostrando la matrice, notiamo delle evidenti similitudini con il punto precedente:

$$ \begin{array}{c|c}
  \color{green}{Q_1}/\color{blue}{Q_0}\color{black}   & Q0 = |0\rangle & Q0 =|1\rangle & \text{Sum} \\ \hline
  Q1 = |0\rangle  & 0 & \color{magenta}{\boxed{49950}}& 49950 \\
  Q1 = |1\rangle & \color{cyan}{\boxed{50050}} & 0 & 50050\\
  \text{Sum} & 50050  & 49950 & 100.000 \\
\end{array} $$

Si differenzi ovviamente col punto precedente, poiché risulta $\text{"diagonale inversa"}$.

Ricollegandoci alle motivazioni date prima, ci chiediamo cosa si noti se andiamo a misurare $\color{blue}{Z_0}$ e poi $\color{green}{Z_1}$.

Supponiamo che la misura su $\color{blue}{Q_0}$ restituisca 1, ovvero collassi nello stato $\color{blue}{|{0}\rangle}$.

$$ \frac{1}{\sqrt2}\big(\color{blue}{|0\rangle}\color{black} \otimes \color{green}{|1\rangle}\color{black}-\color{blue}{|1\rangle}\color{black}\otimes \color{green}{|0\rangle\color{black}\big) \ \ \ \overset {Z_0} \longrightarrow \ \ \ \color{blue}{|0\rangle}\color{black} \otimes \color{green}{|1\rangle}\color{black}  \text{ con }  P = (\frac{1}{\sqrt2})^2 = 0.5$$

Una successiva misura su $\color{green}{Z_1}$ cosa fa?

$$ -\color{blue}{|0\rangle}\color{black} \otimes \color{green}{|1\rangle}\color{black} \ \ \ \overset {Z_1} \longrightarrow \ \ \ \color{blue}{|0\rangle}\color{black} \otimes \color{green}{|1\rangle}\color{black} \text{ con }  P = 1^2 = 1 $$

Il caso in cui $\color{blue}{Z_0}$ faccia collassare lo stato in 

$$ - \color{blue}{|1\rangle}\color{black} \otimes \color{green}{|0\rangle}\color{black}  \text{ con }  P = (\frac{1}{\sqrt2})^2 = 0.5$$ 

si avrá nuovamente determinismo su $\color{green}{Z_1}$, con 

$$ \color{blue}{|1\rangle}\color{black} \otimes \color{green}{|0\rangle}\color{black} \text{ con }  P = (-1)^2 = 1 $$ 

Si noti come il fattore moltiplicativo $-1$ sia una $\text{fase globale}$, non osservabile per definizione, e infatti rimuovibile, in quanto non intacca i calcoli.

Misurando prima $\color{green}{Z_1}$ e poi $\color{blue}{Z_0}$, lo studio non cambia, secondo i ragionamenti mostrati prima.

Quindi anche in questo caso possiamo notare una $perfetta$ correlazione tra le misure sui $\color{blue}{Q_0}$ e $\color{green}{Q_1}$.

In particolare, per lo stato $\color{orange}{|{\Psi^{-}}\rangle}$, abbiamo garanzia che l' osservabile composto $\color{green}{Z_0}\color{black} \otimes \color{blue}{Z_1}$ possa dare come risultato solamente
- $\color{blue}{+1}\color{black},\color{green}{-1}\color{black} \text{ con } P = 0.5  \text{ cioé }  \color{blue}{|1\rangle}\color{black} \otimes \color{green}{|0\rangle} $
- $\color{blue}{-1}\color{black},\color{green}{+1}\color{black} \text{ con } P = 0.5  \text{ cioé } \color{blue}{|0\rangle}\color{black} \otimes \color{green}{|1\rangle} $

Questa informazione é $estremamemente$ utile e possiede un sacco di implicazioni pratiche, in quanto, dopo una misura su $\color{blue}{Q_0}$, permette $istantaneamente$ di poter predirre la misura su $\color{green}{Q_1}$, nonostante non sia ancora stata effettuta.

Nella nostra matrice, questo ragionameno equivale a scegliere una colonna, e avere la certezza che solo una delle 2 righe sará possibile.

Tale ragionamento é analogo per lo stato precedente $\color{orange}{|{\Phi^{+}}\rangle}$, in abbiamo garanzia che l' osservabile composto $\color{blue}{Z_0}\color{black} \otimes \color{green}{Z_1}$ possa dare come risultato solamente
- $\color{blue}{+1}\color{black},\color{green}{+1}\color{black} \text{ con } P = 0.5 \text{ cioé }  \color{blue}{|1\rangle}\color{black} \otimes \color{green}{|1\rangle} $
- $\color{blue}{-1}\color{black},\color{green}{-1}\color{black} \text{ con } P = 0.5 \text{ cioé }  \color{blue}{|0\rangle}\color{black} \otimes \color{green}{|0\rangle} $
